### PR TITLE
Regex support for --secret-type

### PR DESCRIPTION
**Pull request by Nuckal777**
Hi,

this is an implementation for 242, because #243 seems a bit inactive. There are two notable changes:
- dots in file names need to be escaped, due to their special meaning in regex patterns
- the read errors metric behaves differently

### CODE REVIEW

File: internal/kubernetes_test.go [hunk 1]
```diff
@@ -349,22 +350,21 @@ func TestKubeInvalidConfig3(t *testing.T) {
...
-		KubeSecretTypes:   []string{"kubernetes.io/tls:tls.crt", "kubernetes.io/tls:tls.key", "kubernetes.io/tls:nil.key"},
+		KubeSecretTypes: []KubeSecretType{
+			{Type: "kubernetes.io/tls", Regexp: regexp.MustCompile(`tls\.crt`)},
+			{Type: "kubernetes.io/tls", Regexp: regexp.MustCompile(`tls\.key`)},
+			{Type: "kubernetes.io/tls", Regexp: regexp.MustCompile(`nil\.key`)},
+		},
 	}, func(m []model.MetricFamily) {
 		metrics := getMetricsForName(m, "x509_read_errors")
-		assert.Equal(t, 0., metrics[0].GetGauge().GetValue())
+		assert.Equal(t, 1., metrics[0].GetGauge().GetValue())
```
Comment (paullaffitte) [inline internal/kubernetes_test.go:367]: Why does the behavior of `getMetricsForName` change? It should stay the same

File: internal/kubernetes_test.go [hunk 2]
```diff
@@ -349,22 +350,21 @@ func TestKubeInvalidConfig3(t *testing.T) {
...
 
 func TestKubeInvalidSecretType(t *testing.T) {
-	testRequestKube(t, &Exporter{
-		KubeIncludeNamespaces: []string{"default"},
-		KubeSecretTypes:       []string{"aze"},
-	}, func(m []model.MetricFamily) {
-		metrics := getMetricsForName(m, "x509_read_errors")
-		assert.Equal(t, 1., metrics[0].GetGauge().GetValue())
-	})
+	_, err := ParseSecretType("aze")
```
Comment (paullaffitte) [inline internal/kubernetes_test.go:353]: Why did you change the way the test was written? If you want to test in another way, I would prefer that you add a testcase rather than changing the existing when introducing a new feature.

File: internal/exporter.go [hunk 3]
```diff
@@ -52,6 +53,33 @@ type Exporter struct {
...
+	if err != nil {
+		return KubeSecretType{}, err
+	}
+	return KubeSecretType{
+		Type:   ty,
+		Regexp: compiled,
+	}, nil
+}
+
+func (kst *KubeSecretType) Matches(secretType, key string) bool {
```
Comment (paullaffitte) [inline internal/exporter.go:76]: I like the way you implemented this as a struct. Could it be possible to add the following function? The goal would be to use it in places where `Matches` is used inside a loop, it would make the code more readable by decreasing the number of nested loops. (warning: I did not tested neither compiled it).

```golang
func (kst *KubeSecretType) FindMatchingKey(secret *v1.Secret, secretType string, key string) string {
	for key := range secret.Data {
		if secretType.Matches(string(secret.Type), key) {
			return key
		}
	}

	return ""
}
```

**Review by paullaffitte — CHANGES_REQUESTED** [review]
Looks good, however I'm really not comfortable with the fact that you changed the tests without any explanation. I want to be sure that this PR don't introduce any breaking change and those changes suggest that it does.

File: internal/kubernetes_test.go (hunk 1 above)
Comment (Nuckal777) [reply internal/kubernetes_test.go:367]: The current implementation removed the length check ` len(secret.Data[typeAndKey[1]]) > 0 ` on secret values. So, zero length values are returned from `parseAllKubeSecrets()`, which in turn cause parse errors. I could add these checks again.

File: internal/exporter.go (hunk 3 above)
Comment (Nuckal777) [reply internal/exporter.go:76]: The function you outlined would only return the first key matching a regular expression. Some iterator with a callback could be handy though, e.g.:
```go
func IterateMatching(types []KubeSecretType, secret *v1.Secret, cb func(string)) {
	for _, secretType := range types {
		for key := range secret.Data {
			if secretType.Matches(string(secret.Type), key) {
				cb(key)
			}
		}
	}
}
```
That iteration could also deduplicate keys, when a key is matched by multiple regular expressions.

File: internal/kubernetes_test.go (hunk 2 above)
Comment (paullaffitte) [reply internal/kubernetes_test.go:353]: I see, indeed it is the way to go :+1:

**Comment by Nuckal777** [comment]
Hey, what should I follow-up on?

**Comment by Nuckal777** [comment]
Let's try to revive this PR. I repaired the read errors metric.

**Comment by coltaa** [comment]
Any updates on when this could be merged/released?

**Comment by hypnoglow** [comment]
Hi @paullaffitte 

Could we proceed forward with this change?

Our team also eagers for this feature as it is impossible for us to rely on a static list of secret keys.

**Comment by plaffitt** [comment]
Hi @Nuckal777, sorry for the delay. With the latest changes it looks good to me. If you can rebase the branch to resolve conflicts I will merge it. And it would be awesome if you could implement the idea described [here](https://github.com/enix/x509-certificate-exporter/pull/287#discussion_r1604585101) also.

**Comment by Nuckal777** [comment]
Here you go. Because we have go1.23, we can avoid the callback stuff by using iterators. 🎉
